WordPress/gutenberg
# Components: limit `Dropdown` cleanup to renders where the menu has actually been opened

## What?
Followup to https://github.com/WordPress/gutenberg/pull/41505#discussion_r888077070

This PR ensures that the `Dropdown` component doesn't run its cleanup (i.e, close the dropdown) unless the dropdown is actually open.

## How?
Add a check against the `isOpen` prop, which is consistent with the component behavior [prior to a previous refactor](https://github.com/WordPress/gutenberg/pull/23142/files#diff-dacdfb5a25d27771d335dc7307b037ce12d6b122b40b24d1800bd9bd0cd045f7L34-L36)

## Testing Instructions
- Test the `Dropdown` component in Storybook
- From your local Gutenberg directory, run `npx eslint --rule 'react-hooks/exhaustive-deps: warn' packages/components/src/dropdown` to ensure we're staying compliant there
- Confirm unit tests still pass
